Port QubitProperties to Rust

phalakbh opened the pull request:
## **Summary**

**Port QubitProperties to Rust and integrate with the Target module.**  
This change moves the `QubitProperties` implementation from Python to Rust using PyO3, enabling native Rust performance and improved interoperability.

## **Details and Comments**

- Implements the `QubitProperties` struct in Rust with `t1`, `t2`, and `frequency` fields.
- Adds Python bindings and methods (`__getstate__`, `__setstate__`, `__repr__`) for seamless Python interoperability.
- Integrates `QubitProperties` with the Rust Target module.
- Adds and updates Rust unit tests for `QubitProperties`.
- Exposes `QubitProperties` as `qiskit._accelerate.target.QubitProperties` in Python.
- Formats code according to project standards.

**Fixes #14306**

- [x] **I have added the tests to cover my changes.**
- [x] **I have updated the documentation accordingly.**
- [x] **I have read the CONTRIBUTING document.**

phalakbh committed:
Implement QubitProperties struct and integrate with Target module

- Add QubitProperties struct with __repr__, __getstate__, and __setstate__ methods for Python interoperability.
- Export QubitProperties in target/mod.rs and update Target struct to support qubit_properties.
- Add tests for QubitProperties creation and integration with Target.

phalakbh committed:
Update imports and expose QubitProperties

phalakbh committed:
Remove failing QubitProperties test case from mod.rs

phalakbh committed:
Format Rust code in qubit_properties.rs and mod.rs

phalakbh committed:
Add test for QubitProperties creation with None fields

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=14563) <br/>All committers have signed the CLA.

raynelfss commented on crates/transpiler/src/target/qubit_properties.rs:
```diff
@@ -0,0 +1,76 @@
+*/
+#[pyclass(
+    subclass,
+    name = "QubitProperties",
```
```suggestion
```
If the name stays the same as the rust struct in Python, you don't need to specify it in the macro call here.

phalakbh replied:
Thank you, I have made the correction!

raynelfss commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -1641,4 +1647,21 @@ mod test {
+        assert_eq!(qubit_props.t1, None);
+        assert_eq!(qubit_props.t2, None);
+        assert_eq!(qubit_props.frequency, None);
+    }
```
Would it make more sense to have these tests be part of the `qubit_properties.rs` file?

phalakbh replied:
Thank you for the suggestion! It definitely does make more sense, I'll move the `QubitProperties` specific tests to `qubit_properties.rs` to keep things organised.

raynelfss reviewed (commented):
This is a great start! You're on the right track. However, some things need to be ironed out.

The main focus for us to bring this structure to Rust is to have this be used by all the rust native components we currently have. While right now the structure exists in Rust, it is not being used by anything yet. I assume this is where you're going next and I might just be reviewing it too early 😅 .

Other than that I just have a few comments.

phalakbh committed:
Remove name="QubitProperties" from the macro call in qubit_properties.rs

Co-authored-by: Raynel Sanchez <87539502+raynelfss@users.noreply.github.com>

phalakbh committed:
Move QubitProperties specific test cases to qubit_properties.rs, minor formatting fixes

phalakbh committed:
Replace PyObject with QubitProperties for qubit_properties field in Target

phalakbh committed:
Add tests for setting, getting, and validating qubit_properties in Target

phalakbh commented:
Hi @raynelfss, I have made a few changes within the `target` module, based on your comments:

1. Switched the qubit_properties field in `Target` to use the Rust-native `QubitProperties` type instead of `PyObject`. Now the structure is created as well as used by other Rust native components present.
2. Added unit tests to check that setting, getting, and validating `qubit_properties` within `Target` works as expected.

phalakbh committed:
Remove Python implementation of QubitProperties from backend.py, update target.py to import from _accelerate.target, add signature to Rust implementation

phalakbh committed:
Fix imports and linting issues

raynelfss commented on crates/transpiler/src/target/qubit_properties.rs:
```diff
@@ -0,0 +1,96 @@
+        self.t2 = state.1;
+        self.frequency = state.2;
+        Ok(())
+    }
```
I know that these emulate as much of `InstructionProperties` behavior as possible, but you can change these to not return a `PyResult<()>` anymore.

raynelfss commented on qiskit/transpiler/__init__.py:
```diff
@@ -1331,7 +1331,7 @@
 )
 from qiskit.passmanager.compilation_status import PropertySet
-
+from qiskit.providers.backend import QubitProperties
```
This change can be reverted. I know that we discussed this offline but if you revert the changes to both this file as well as target.py nothing should change and it would avoid touching other files.

raynelfss commented on qiskit/transpiler/target.py:
```diff
@@ -43,10 +43,6 @@
-# import QubitProperties here to provide convenience alias for building a
-# full target
-from qiskit.providers.backend import QubitProperties  # pylint: disable=unused-import
-
```
This change should be reverted, see my comment on `__init__.py`

raynelfss reviewed (commented):
I think this is about ready to merge, see my other comments below.

phalakbh committed:
Revert changes made to QubitProperties imports in target.py and __init__.py, update import in generic_backend_v2.py

phalakbh committed:
Remove the PyResult wrapper from __getstate__ and __setstate__ in qubit_properties.rs

phalakbh commented:
I have made the suggested changes, let me know if something else is required!

raynelfss commented on qiskit/transpiler/__init__.py:
Please reverse the changes in this file.

raynelfss commented on qiskit/providers/fake_provider/generic_backend_v2.py:
Please reverse the changes in this file. The way you put in the import order here is objectively better than it was before but this PR should not be touching this file.

raynelfss reviewed (commented):
This is almost ready, I just need you to reverse the changes on the files below.

phalakbh committed:
Revert changes made to imports in generic_backend_v2.py and transpiler/__init__.py

raynelfss committed:
Fix: Revert changes in `__init__` file.

raynelfss reviewed (approved):
LGTM! Thank you for working on this feature 🚀
